docs/spec_draft.md

```diff
@@ -343,6 +344,120 @@ logical operation.
+
+```python
+def reduce_sum(operand, feature_index):
+  (sum,) = reduce(
+      inputs=[operand],
+      init_values=[0],
+      dimensions=[i for i in range(rank(operand)) if i != feature_index],
+      body=lambda x, y: add(x, y))
+  return sum
+
+def compute_mean(operand, feature_index):
+  sum = reduce_sum(operand, feature_index)
+  divisor = constant(num_elements(operand) / dim(operand, feature_index))
+  divisor_bcast = broadcast_in_dim(divisor, [], shape(sum))
+  return divide(sum, divisor_bcast)
+
+def batch_norm_grad(operand, scale, mean, variance, grad_output, epsilon, feature_index):
+  # Broadcast inputs to shape(operand)
+  scale_bcast = broadcast_in_dim(scale, [feature_index], shape(operand))
+  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
+  variance_bcast = broadcast_in_dim(variance, [feature_index], shape(operand))
+  epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))
+
+  centered_operand = subtract(operand, mean_bcast)
+  stddev = sqrt(add(variance_bcast, epsilon_bcast))
+  normalized_operand = divide(centered_operand, stddev)
+
+  intermediate_cl = compute_mean(multiply(grad_output, divide(normalized_operand, stddev)))
+  mean_grad_output = compute_mean(grad_output, feature_index)
+
+  intermediate_cl_bcast = broadcast_in_dim(intermediate_cl, [feature_index], shape(operand))
+  mean_grad_output_bcast = broadcast_in_dim(mean_grad_output, [feature_index], shape(operand))
+
+  grad_operand = multiply(
+    divide(scale_bcast, stddev),
+    subtract(
+      subtract(grad_output, mean_grad_output_bcast),
+      multiply(intermediate_cl_bcast, centered_operand)
+    )
+  )
+  grad_scale = reduce_sum(multiply(grad_output, normalized_operand), feature_index)
+  grad_offset = reduce_sum(grad_output, feature_index)
+  return grad_operand, grad_scale, grad_offset
+```
+
+### Inputs
+
+| Name            | Type                                        |
+|-----------------|---------------------------------------------|
+| `operand`       | tensor of floating-point type               |
+| `scale`         | 1-dimensional tensor of floating-point type |
+| `mean`          | 1-dimensional tensor of floating-point type |
+| `variance`      | 1-dimensional tensor of floating-point type |
+| `grad_output`   | tensor of floating-point type               |
+| `epsilon`       | constant of type `f32`                      |
+| `feature_index` | constant of type `si64`                     |
+
+### Outputs
+
+| Name           | Type                                        |
+|----------------|---------------------------------------------|
+| `grad_operand` | tensor of floating-point type               |
+| `grad_scale`   | 1-dimensional tensor of floating-point type |
+| `grad_offset`  | 1-dimensional tensor of floating-point type |
+
+### Constraints
+
+  * (C1) 0 $\le$ `feature_index` $\lt$ rank(`operand`).
+  * (C2) `operand`, `scale`, `mean`, `variance`, `grad_output`, `grad_operand`
```

Comment:
Drop (C2) as already mentioned in the tables above?

Reply:
Tables only say they are of floating type, this is stricter that that, it says they should match exactly.